### PR TITLE
feat: add the next event's info to the home page

### PR DESCRIPTION
**Pull request by colbywhite**
## Description
As is, the site doesn't have any functionality that leverages remix too much since it's a static site. In the hopes of having an example of SSR will still also providing some value, I've used the Meetup API in order to display the group's next listed event on the home page during the SSR. If there is no listed event then is displays the previous generic button. 

### Screenshots

| mobile | desktop|
|---|---|
| Screenshot 2022-12-04 at 17-23-58 Remix Austin 💿 A community   monthly Meetup event for Remix developers | Screenshot 2022-12-04 at 17-24-37 Remix Austin 💿 A community   monthly Meetup event for Remix developers

### Notes
- I avoided testing pure presentation components (those type of tests are prone to rot) and focused on tests around functionality. if you want even deeper testing, i'll add
- the [meetup api](https://www.meetup.com/api/schema/#groupByUrlname) is open to the public and doesn't require auth
- some of the components on the index route were refactored into separate components to make the index route slimmer and more focused (and thus easier to reason about how the loader and the comp interact).
- The styling of the info is very straightforward. happy to make it fancier if someone has some ideas in a follow-up PR.

### CODE REVIEW

**Comment by brookslybrand** [comment]
Awesome, thank you you @colbywhite! I love the idea of using Meetup's API to both leverage Remix's SSR features, as well as making the call to action a little more exciting.

I also appreciate you doing some code cleanup/refactoring. Given that this PR is a little larger than the required change (due to the refactors), would it be possible to split this PR into 2 PRs? It would make reviewing the changes a little easier in my opinion.

Would also love for the button to not look so elongated on mobile, so yes, definitely looking for others feedback. To start I think we can drop the word "come".

**Comment by colbywhite** [comment]
@brookslybrand, good call out. i tried to keep the `refactor` commit and `feat` commit isolated, but got a little carried away. refactor stuff is removed and i'll put it in a diff pr later. 

as for mobile, i shortened the text by a word which helped. but any better styling ideas are welcome. i could also drop the idea of putting the amount of people already reserved in the button text.

| mobile |
|---|
|Screenshot 2022-12-05 at 14-10-28 Remix Austin 💿 A community   monthly Meetup event for Remix developers|

**Comment by brookslybrand** [comment]
Awesome! Thanks Colby, sorry for the delays, going to do a review now

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1 @@
+*  text=auto eol=lf
```
Comment (brookslybrand) [inline .gitattributes:1]: Curious what this is for, I'm unfamiliar

File: app/components/MeetupButton.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import MeetupIcon from "~/components/icons/MeetupIcon";
+import type { PropsWithChildren } from "react";
+
+export default function MeetupButton({
+  link,
+  children,
+}: PropsWithChildren<{ link: string }>) {
```
Comment (brookslybrand) [inline app/components/MeetupButton.tsx:7]: Nit: I prefer simply:

```
{
  link: string;
  children: React.ReactNode
}
```

Mostly since the child is actually required.

File: app/components/NextEventInfo/NextEventInfo.test.tsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      "Fri. Dec 31, 7:00 PM CST"
+    );
+  });
+
+  it("renders button label when many are attending", () => {
+    const eventWithManyAttending = { ...DEFAULT_ONLINE_EVENT, going: 10 };
+    render(<NextEventInfo event={eventWithManyAttending} />);
+    expect(screen.getByTestId("meetup-button-text")).toHaveTextContent(
+      "Join 10 others at our next meetup!"
+    );
```
Comment (brookslybrand) [inline app/components/NextEventInfo/NextEventInfo.test.tsx:60]: Do you mind trying to use test ID's a little less? You can achieve the same thing and even add a little bit more value to the test with a `getByRole` query:

```ts
screen.getByRole("link", { name: "Join 10 others at our next meetup!" })
```

The creator of the [testing-library](https://testing-library.com/) framework we're using (on top of vitest) has some good advice/reasoning behind this: https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#not-querying-by-text

In short:
- Less setup/maintenance keeping track of all the test IDs
- Better accessibility testing when querying by roles/text

File: app/utils.ts [hunk 4]
```diff
@@ -45,3 +45,7 @@ export function useMatchesData(
 export function validateEmail(email: unknown): email is string {
   return typeof email === "string" && email.length > 3 && email.includes("@");
 }
+
+export function isEmptyString(val: string): boolean {
+  return val !== undefined && val.trim().length <= 0;
```
Comment (brookslybrand) [inline app/utils.ts:50]: Out of curiosity, why test if val is not undefined if val can only be a string?

File: app/components/NextEventInfo/NextEventInfo.tsx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+function isUndefinedVenue(venue: MeetupEvent["venue"]): boolean {
+  return (
+    isEmptyString(venue.address) ||
+    isEmptyString(venue.city) ||
+    isEmptyString(venue.state)
+  );
+}
+
+export default function NextEventInfo({ event }: { event: MeetupEvent }) {
+  const date = DateTime.fromISO(event.dateTime).setZone(CENTRAL_TIMEZONE);
```
Comment (brookslybrand) [inline app/components/NextEventInfo/NextEventInfo.tsx:24]: Do you mind using [`Intl.DateTimeFormat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat) since it's built into the browser?

Luxon is a [pretty big](https://bundlephobia.com/package/luxon@3.1.1) library just for this one utility (it may tree shake well, I'm not sure)

File: app/components/MeetupButton.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import MeetupIcon from "~/components/icons/MeetupIcon";
+import type { PropsWithChildren } from "react";
+
+export default function MeetupButton({
```
Comment (brookslybrand) [inline app/components/MeetupButton.tsx:4]: Mind renaming this to `MeetupLink` since this is not a button?

File: app/components/NextEventInfo/NextEventInfo.tsx (hunk 5 above)
Comment (brookslybrand) [reply app/components/NextEventInfo/NextEventInfo.tsx:24]: Also, I don't think we need to set the timezone, since the ISO string should have it built in and we want to display to the users the time in their timezone. Perhaps I'm not thinking about this correctly though

File: app/models/meetup.server.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+  const body: MeetupRequestBody = {
+    variables: { urlname: url },
+    query: GET_GROUP_QUERY,
+  };
+  return fetch(MEETUP_API, {
+    method: "POST",
+    body: JSON.stringify(body),
+    headers: JSON_HEADERS,
+  }).then((res) => res.json() as Promise<MeetupGroupResponse>);
+}
```
Comment (brookslybrand) [inline app/models/meetup.server.ts:45]: We should probably parse to make sure this value is correct instead of type casting. Additionally, do we have any error handling?

Also, as I'm looking at this, if you want to use luxon maybe do it here in the server, that way we won't be sending the entire package just to massage a single field.

Definitely don't hesitate to message this data into the shape you want before sending it down the loader, no need to lots of work in the component when it can be done at the loader level.

File: app/models/meetup.server.test.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+      server.use(
+        rest.post("https://www.meetup.com/gql", async (req, res, context) => {
+          requestBody = await req.json();
+          const response: MeetupGroupResponse = {
+            data: { groupByUrlname: MOCK_GROUP },
+          };
+          return res(context.status(200), context.json(response));
+        })
+      );
+    });
```
Comment (brookslybrand) [inline app/models/meetup.server.test.ts:48]: Nice! I like seeing the use of msw for the mock server.

Here might be a good place to add some tests for api failures.

File: app/models/meetup.d.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
+export interface MeetupEvent {
```
Comment (brookslybrand) [inline app/models/meetup.d.ts:1]: Thanks for adding all of these types and so much organization! Mind just making this `meetup.ts`? We don't really need it to be a type definition file and I think it could just cause confusion down the line for other devs.

**Review by brookslybrand — CHANGES_REQUESTED** [review]
Hey @colbywhite, great work! Added a number of comments, mostly around code style and organization, please let me know if you have any question!

Definitely realizing we might need to add a little bit more info in our contributing guide around best practices, and maybe a doc on code organization. The cool thing is your helping shape all that right now!

File: test/setup-test-env.ts [hunk 10]
```diff
@@ -1,15 +1,26 @@
 import { installGlobals } from "@remix-run/node";
 import "@testing-library/jest-dom/extend-expect";
-import { expect, afterEach } from "vitest";
+import { afterEach, beforeAll, expect } from "vitest";
 import { cleanup } from "@testing-library/react";
 import matchers from "@testing-library/jest-dom/matchers";
+import { server } from "../mocks/server";
```
Comment (grantnorwood) [inline test/setup-test-env.ts:6]: Similar to above comment ... can we add an `export` to the server object defined `mocks/index.ts` to reuse that instead of creating a new `server.ts`

```ts
export const server = setupServer();
```

File: app/models/meetup.server.test.ts (hunk 8 above)
Comment (grantnorwood) [reply app/models/meetup.server.test.ts:48]: If you happen to be able to reuse [`mocks/index.js`](https://github.com/remix-austin/remixaustin-com/blob/main/mocks/index.js) _(may have to make it a `.ts`)_, would you mind adding this test to it? 🙂

**Comment by grantnorwood** [comment]
@colbywhite This is looking great!  Left a couple comments on having two server mocks.

Also, would you mind adding some spacing above the "Join 17 others ..." button?  Just looks a little cramped.

image

File: .gitattributes (hunk 1 above)
Comment (colbywhite) [reply .gitattributes:1]: [ensures that windows' users don't accidentally check in the wrong line endings](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings#per-repository-settings).

File: app/models/meetup.server.test.ts (hunk 8 above)
Comment (colbywhite) [reply app/models/meetup.server.test.ts]: i think doing so would affect `npm run test:e2e:run`. those smoke tests expect requiring the `./mocks` to have the side effect of running `server.listen()` in a separate process. however, for these tests i'm doing the standard msw strategy of hitting that in a `beforeAll`. since i'm not too familiar with cypress i decided not to mess with it and stick to the standard msw strategy of doing this in `./mocks/server`.

File: app/models/meetup.d.ts (hunk 9 above)
Comment (colbywhite) [reply app/models/meetup.d.ts]: that was on purpose in order to not cause confusion. a `.d.ts` indicates there's only types in this file. however, this point is moot simply b/c the process of adding yup-based parsing added more than types to this file so i renamed it accordingly.

File: app/components/NextEventInfo/NextEventInfo.tsx (hunk 5 above)
Comment (colbywhite) [reply app/components/NextEventInfo/NextEventInfo.tsx]: since we're doing SSR, the user's timezone becomes tricky. when we SSR, the timezone will be UTC. the browser will be something else. and thus you now have to deal with the blip: SSR render in UTC but the hydration shifts it to a diff time zone. (i also think that react will give you a warning/error that the content is different. but it's a bit annoying since you won't see it during local dev.) 

[there are solutions](https://edgebound.dev/the-pre-rendering-blip). likely rendering a default timezone via SSR and let the client do a CSS animation into the browser's time zone.

on a related but different note, i think a physical, in-person event is a scenario where it makes sense to _not_ show the time in the browser's timezone. and use the event's timezone.

and thus i elected to forgo the blip altogether and render in the event's timezone.

File: app/utils.ts (hunk 4 above)
Comment (colbywhite) [reply app/utils.ts]: i think the real answer here is habit 😅 . i'll make this a more generic function.

File: app/components/MeetupButton.tsx (hunk 2 above)
Comment (colbywhite) [reply app/components/MeetupButton.tsx]: interesting. i don't think i ever realized that was optional. i'd argue `PropsWithChildren` is a misleading name then, but that's neither here nor there. 😉 

fixed via:
```typescript
type PropsWithRequiredChildren<P = unknown> = P & {children: ReactNode};
```

File: app/models/meetup.server.test.ts (hunk 8 above)
Comment (colbywhite) [reply app/models/meetup.server.test.ts]: update: i didn't notice that cypress was swapped for playwright. doing so would still affect `npm run dev` in a similar fashion due to the usage of `binode --require ./mocks -- @remix-run/dev:remix dev`.

investigating more, the indie stack puts `mocks/index.js` there in order to allow for you to mock something during local dev. seems like a remix convention. (or maybe a kcd convention). so mixing the `beforeAll` stuff into the same file isn't ideal. so the test msw and this dev msw need to be different.

File: .gitattributes (hunk 1 above)
Comment (brookslybrand) [reply .gitattributes:1]: Ah great! Thanks for explaining 🙂

File: app/components/MeetupButton.tsx (hunk 2 above)
Comment (brookslybrand) [reply app/components/MeetupButton.tsx]: Haha I would agree with that argument. I really don't like the default `children` as optional thing, since most custom components that take in children really want them to be required.

Thanks Colby!

File: app/models/meetup.d.ts (hunk 9 above)
Comment (brookslybrand) [reply app/models/meetup.d.ts]: That makes sense. Yeah, I think my thought process is just that I often I'll end up adding more than just types to a file that starts as only types, so starting with `.ts` is one less thing to think about in the future.

Also, types declared in `.d.ts` files makes the types global, which is generally not what we want to do
image

**Comment by brookslybrand** [comment]
Extra spacing looks _much better_ @colbywhite!

Can we make sure the spacing is the same amount on the bottom as well though?

image

File: app/components/NextEventInfo/NextEventInfo.tsx (hunk 5 above)
Comment (brookslybrand) [reply app/components/NextEventInfo/NextEventInfo.tsx]: Nice self plug 👀 

Right now I'd rather avoid doing any client-side hydration, since we really don't need JS (yet) #40 

> on a related but different note, i think a physical, in-person event is a scenario where it makes sense to not show the time in the browser's timezone. and use the event's timezone.

I think you're totally right, which is what I came here to say. Just saying "CST" no matter the user's time zone seems like the right way to go, so I'm on board with your solution 💪

File: app/components/NextEventInfo/NextEventInfo.tsx [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  timeZone: "America/Chicago",
+});
+
+function isUndefinedVenue(venue: MeetupEvent["venue"]): boolean {
+  return (
+    isEmptyString(venue.address) ||
+    isEmptyString(venue.city) ||
+    isEmptyString(venue.state)
+  );
+}
```
Comment (brookslybrand) [inline app/components/NextEventInfo/NextEventInfo.tsx:29]: I don't think the logic of `isEmptyString` is quite right, because `isUndefinedVenue({})` will return `false`, when it should return `true`.

Might want to add a test to `NextEventInfo.test.tsx` to check for missing values, not just values with empty strings

File: app/models/meetup.parsing.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  shortUrl: string().required().url(),
+  venue: object({
+    name: string().required(),
+    address: string().defined().strict(true),
+    city: string().defined().strict(true),
+    state: string().defined().strict(true),
+  }).required(),
+  dateTime: date().required(),
+  going: number().required(),
+}).required();
```
Comment (brookslybrand) [inline app/models/meetup.parsing.ts:15]: Thanks so much for adding parsing/a schema! I'm a little partial to `zod`, but since you're the one doing the work I'm happy with this 😄 

It's now on me to give a good reason to use `zod` instead in the future, so I'll definitely tag you if I get around to that

File: app/models/meetup.server.test.ts (hunk 8 above)
Comment (brookslybrand) [reply app/models/meetup.server.test.ts]: > seems like a remix convention. (or maybe a kcd convention).

😅 I think lots of the Remix testing conventions are kcd things

I'm okay having them separate for now, but it does seem a bit strange to have 2 versions of the same thing.

I guess the only problem using the same msw server for both is if you were running `dev` and `test` at the same time, there could be clashes 🤷‍♂️ 

Anyway, I'm not too concerned, especially since we're not really using the mocks much (yet)

File: test/setup-test-env.ts [hunk 13]
```diff
@@ -1,15 +1,26 @@
...
 // extends Vitest's expect method with methods from react-testing-library
 expect.extend(matchers);
 
+const closeServer = () => server.close();
+
+beforeAll(() => {
+  server.listen({ onUnhandledRequest: "bypass" });
+  process.once("SIGINT", closeServer);
+  process.once("SIGTERM", closeServer);
+  return closeServer;
```
Comment (brookslybrand) [inline test/setup-test-env.ts:19]: I'm a little confused why we're closing the server immediately after starting it? This gets run right after this hook and before all tests from what I can tell.

Comment (colbywhite) [reply test/setup-test-env.ts:19]: [the returned function of `beforeAll` is essentially the `afterAll`.](https://vitest.dev/api/#beforeall)

File: app/models/meetup.d.ts (hunk 9 above)
Comment (colbywhite) [reply app/models/meetup.d.ts]: huh. learn something new everyday. wasn't aware of the global side effect.
> If you want an interface/type that you can simply use without having to import it in the consuming module said interface must reside in a .ts or ideally a .d.ts file **_without any imports or exports being present in the same file._** This is of utmost importance because as soon as you are exporting at least one thing from the same file it becomes a module and everything that is in that module must be subsequently imported by consumers.

https://stackoverflow.com/q/42233987

so my previous `.d.ts` would've been a module since i was exporting things and thus no global types. but admittedly, that was on accident. 😅 such a unintuitive behavior choice from TS. now that i know that, i'll probably use `.d.ts` far less often. 

but yea, the parsing made this point moot. but thanks for sending me down that rabbit hole on odd TS behavior.

File: app/components/NextEventInfo/NextEventInfo.tsx (hunk 11 above)
Comment (colbywhite) [reply app/components/NextEventInfo/NextEventInfo.tsx:29]: interesting. `{}` isn't a valid scenario due to the yup parsing. in order to build a piece of data to test, you'd have to force a cast just to build it. thus it feels a bit overkill to handle that scenario.

File: test/setup-test-env.ts (hunk 13 above)
Comment (brookslybrand) [reply test/setup-test-env.ts:19]: Ah okay, gotcha. I did a manual test and it seemed like it called the return function before it ended. No worries though! I see what you're saying based on the documentation you shared.

**Comment by brookslybrand** [comment]
@colbywhite I think once the bottom spacing is added I'm good to approve! Thanks so much for working on this, really great stuff

**Review by brookslybrand — APPROVED** [review]
Looks great! Thanks for all the hard work @colbywhite 🙂 

Go ahead and merge when you're ready

**Comment by colbywhite** [comment]
@brookslybrand , fyi, due to perms it looks like one of yall will have to merge.

**Comment by brookslybrand** [comment]
> @brookslybrand , fyi, due to perms it looks like one of yall will have to merge.

Ah yes, that’s my bad. Merged!

**Comment by grantnorwood** [comment]
@colbywhite Thanks so much for this PR!  This is really neat, and makes our site feel more dynamic & connected.
